reco_utils/recommender/rbm/rbm.py

```diff
@@ -0,0 +1,946 @@
+
+        else:
+            answer = tm.time() - self.start_time
+            # reset state
+            self.start_time = None
+            return answer
+
+        #else:
+        #    return None
+
+    #Binomial sampling
+    def B_sampling(self,pr):
+
+        '''
+        Binomial sampling of hidden units activations using a rejection method.
+
+        Args:
+            pr: input conditional probability
+            g:  uniform probability used for comparison
+
+        Returns:
+            h_sampled: sampled units. The value is 1 if pr>g and 0 otherwise.
+
+        Basic mechanics:
+            1) Extract a random number from a uniform distribution (g) and compare it with
+                the unit's probability (pr)
+
+            2) Choose 0 if pr<g, 1 otherwise. It is convenient to implement this condtion using
+               the relu function.
+
+        '''
+
+        np.random.seed(1)
+
+        #sample from a Bernoulli distribution with same dimensions as input distribution
+        g = tf.convert_to_tensor(np.random.uniform(size= pr.shape[1] ), dtype=tf.float32)
+
+        #sample the value of the hidden units
+        h_sampled = tf.nn.relu(tf.sign(pr-g) )
+
+        return h_sampled
+
+    #Multinomial sampling
+    def M_sampling(self, pr):
+
+        '''
+        Multinomial Sampling of ratings
+
+        Args:
+            pr:  a distributions of shape (m, n, r), where m is the number of examples, n the number
+                 of features and r the number of classes. pr needs to be normalized, i.e.
+                 sum_k p(k) = 1 for all m, at fixed n.
+            f:   normalized, uniform probability used for comparison.
+
+        Returns:
+            v_samp: an (m,n) array of sampled rankings from 1 to r .
+
+        Basic mechanics:
+                For r classes, we sample r binomial distributions using the rejection method. This is possible
+                since each class is statistically independent from the other. Note that this is the same method
+                used in numpy's random.multinomial() function.
+
+                1) extract a size r array of random numbers from a uniform distribution (g). As pr is normalized,
+                   we need to normalize g as well.
+
+                2) For each user and item, compare pr with the reference distribution. Note that the latter needs
+                   to be the same for ALL the user/item pairs in the dataset, as by assumptions they are sampled
+                   from a common distribution.
+
+        '''
+
+        np.random.seed(1)
+
+        g = np.random.uniform(size= pr.shape[2]) #sample from a uniform distribution
+        f = tf.convert_to_tensor(g/g.sum(), dtype= tf.float32 )#normalize and convert to tensor
+
+        samp = tf.nn.relu(tf.sign(pr - f) ) #apply rejection method
+        v_samp = tf.cast( tf.argmax(samp , axis=2)+1, 'float32') #select sampled element
+
+        return v_samp
+
+    #Multinomial distribution
```

Comment:
One example to follow google docstrings would be this:
```
     def Pm(self, phi): 
         """ Multinomial distribution. Probability that unit v has value l given phi: P(v=l|phi).
         Args:
             phi (np.array): Linear... 
```
Not sure whether the input is a np.array or a tf tensor...